client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,455 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package clientutil
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/timeutil"
+)
+
+var (
+	// this list is a "nice" "human" "readable" "ordering" of headers to print
+	// off, sorted in lexical order with meta headers and primary key headers
+	// removed, and big nasty keys such as device-key-sha3-384 and
+	// device-key at the bottom
+	// it also contains both serial and model assertion headers, but we
+	// follow the same code path for both assertion types and some of the
+	// headers are shared between the two, so it still works out correctly
+	niceOrdering = [...]string{
+		"architecture",
+		"base",
+		"classic",
+		"display-name",
+		"gadget",
+		"kernel",
+		"revision",
+		"store",
+		"system-user-authority",
+		"timestamp",
+		"required-snaps", // for uc16 and uc18 models
+		"snaps",          // for uc20 models
+		"device-key-sha3-384",
+		"device-key",
+	}
+)
+
+type ModelFormatter interface {
+	LongPublisher(storeAccountID string) string
+	GetEscapedDash() string
+}
+
+type PrintModelAssertionOptions struct {
+	TermWidth int
+	AbsTime   bool
+	Verbose   bool
+	Assertion bool
+}
+
+func fmtTime(t time.Time, abs bool) string {
+	if abs {
+		return t.Format(time.RFC3339)
+	}
+	return timeutil.Human(t)
+}
+
+func formatInvalidTypeErr(headers ...string) error {
+	return fmt.Errorf("invalid type for %q header", strings.Join(headers, "/"))
+}
+
+func printVerboseSnapsList(w *tabwriter.Writer, snaps []interface{}) error {
+	printModes := func(snapName string, members map[string]interface{}) error {
+		modes, ok := members["modes"]
+		if !ok {
+			return nil
+		}
+
+		modesSlice, ok := modes.([]interface{})
```

Comment:
does it work if you try `modes.([]string)` ?

Reply:
It works for my current tests - however I'm a bit hesitant to change this as this does not match current behaviour, because this would mean that we do not print any modes if one of them is not string, where as the current behaviour would just skip it.